server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java

```diff
@@ -152,29 +176,18 @@ public int getNumberOfSegmentsInQueue()
   @Override
   public void loadSegment(final DataSegment segment, final LoadPeonCallback callback)
   {
-    synchronized (lock) {
-      if ((currentlyProcessing != null) &&
-          currentlyProcessing.getSegmentId().equals(segment.getId())) {
-        if (callback != null) {
-          currentlyProcessing.addCallback(callback);
-        }
-        return;
-      }
-    }
+    SegmentHolder segmentHolder = new SegmentHolder(segment, LOAD, Collections.singletonList(callback));
```

Comment:
Does it make sense that we create a singleton list of null potentially, but a few lines below avoid adding null as a callback?

Reply:
I actually like the way this code reads.
If I end up passing a null element in the singleton list, then downstream code like this
`List<LoadPeonCallback> snapshotCallbacks()
    {
      synchronized (callbacks) {
        // Return an immutable copy so that callers don't have to worry about concurrent modification
        return ImmutableList.copyOf(callbacks);
      }
    }`
would break. It is not worth the change, IMHO.

Reply:
The question is if the downstream code prepared to see nulls in this list? Even if it does, it would be better if it didn't and it was a guarantee that the list never contains nulls. It's less error-prone.

Reply:
Moved null checks downstream to SegmentHolder class.